{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5be0aabb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-16T10:15:43.489438Z",
     "start_time": "2023-01-16T10:15:43.108318Z"
    }
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import matplotlib.pyplot as pl\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import gseapy as gp\n",
    "import scvelo as scv\n",
    "from tqdm.auto import tqdm\n",
    "from pathlib import Path\n",
    "\n",
    "# Paths and utils\n",
    "root = Path('../../')\n",
    "figure_path = root / 'results/figures/Neutrophils/'\n",
    "figure_path.mkdir(parents=True, exist_ok=True)\n",
    "table_path = root / 'results/tables/'\n",
    "signatures_path = root / 'resources/'\n",
    "import yaml\n",
    "with open(root / 'config/config.yaml', 'r') as file:\n",
    "    config = yaml.safe_load(file)\n",
    "data_path = config['data_path']\n",
    "\n",
    "# Set plotting style\n",
    "from matplotlib import rcParams\n",
    "# Type 2/TrueType fonts\n",
    "rcParams['pdf.fonttype'] = 42\n",
    "rcParams['ps.fonttype'] = 42\n",
    "# Arial font\n",
    "rcParams['font.sans-serif'] = \"Arial\"\n",
    "rcParams['font.family'] = \"sans-serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e65ac14f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# utility functions used for analysis\n",
    "\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as pl\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import seaborn as sns\n",
    "import gseapy as gp\n",
    "import statsmodels.api as sm\n",
    "import decoupler as dc\n",
    "\n",
    "from functools import wraps\n",
    "from tqdm.auto import tqdm\n",
    "from warnings import warn\n",
    "from pathlib import Path\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import ttest_ind, zscore\n",
    "from scipy.sparse import issparse\n",
    "from scipy.cluster.hierarchy import distance, linkage, dendrogram\n",
    "\n",
    "signatures_path_ = root / 'resources/'\n",
    "\n",
    "\n",
    "# Newer names, more accurate\n",
    "organism_rename_dict = {\n",
    "    'Dwarfhamster': 'Roborovski hamster',\n",
    "    'Goldhamster': 'Syrian hamster'\n",
    "}\n",
    "time_rename_dict = {\n",
    "    'D0': '0 dpi',\n",
    "    'D2': '2 dpi',\n",
    "    'D3': '3 dpi',\n",
    "    'D5': '5 dpi',\n",
    "    'E14': '14 dpi'\n",
    "}\n",
    "dosage_rename_dict = {\n",
    "    'no dose': '0 pfu',\n",
    "    'low dose': r'$1\\times10^4$ pfu',\n",
    "    'high dose': r'$1\\times10^5$ pfu'\n",
    "}\n",
    "\n",
    "def get_key_values(adata, key):\n",
    "    if key in adata.var_names:\n",
    "        from scipy.sparse import issparse\n",
    "        return adata[:, key].X.A.T[0] if issparse(adata.X) else adata[:, key].X.T[0]\n",
    "    elif key in adata.obs.keys():\n",
    "        return adata.obs[key]\n",
    "    else:\n",
    "        raise ValueError(f'Could not find key {key} in neither obs nor var_names.')\n",
    "\n",
    "def rsum(X, axis):\n",
    "    # handles sparse sum, returns array instead of matrix object\n",
    "    return np.ravel(np.sum(X, axis=axis))\n",
    "\n",
    "def extract_df(adata, keys):\n",
    "    df = pd.DataFrame(index=adata.obs_names)\n",
    "    for key in keys:\n",
    "        df[key] = get_key_values(adata, key)\n",
    "    return df\n",
    "\n",
    "def cluster_matrix(matrix, how='row', return_order=False, method='centroid'):\n",
    "    '''\n",
    "    Hierarchical clustering of a matrix/dataframe. `how` can be 'col', 'row' or 'both' (default: 'row').\n",
    "    '''\n",
    "    if how not in ['col', 'row', 'both']:\n",
    "        raise ValueError('Value for \"how\" must be row or col.')\n",
    "    if how!='both':\n",
    "        M = matrix if how=='row' else matrix.T\n",
    "        dist = distance.pdist(M)\n",
    "        link = linkage(dist, method=method)\n",
    "        dend = dendrogram(link, no_plot=True)\n",
    "        order = np.array(dend['leaves'], dtype=int)\n",
    "        if return_order:\n",
    "            return order\n",
    "        elif isinstance(matrix, pd.DataFrame):\n",
    "            return matrix.iloc[order] if how=='row' else matrix.iloc[:, order]\n",
    "        else:\n",
    "            return matrix[order] if how=='row' else matrix[:, order]\n",
    "    else:\n",
    "        if return_order:\n",
    "            warn('Returning order when clustering both row and col is not supported.')\n",
    "        matrix_ = cluster_matrix(matrix, how='row', return_order=False, method=method)\n",
    "        return cluster_matrix(matrix_, how='col', return_order=False, method=method)\n",
    "\n",
    "def format_one_pval(val, star='*'):\n",
    "    \"\"\"\n",
    "    Format p-value for display in a table or plot.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    val : float\n",
    "        p-value.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    formatted : str\n",
    "        Formatted p-value.\n",
    "    \"\"\"\n",
    "    if val > 0.05:\n",
    "        return 'ns'\n",
    "    elif val < 0.0001:\n",
    "        return star*4  # '****'\n",
    "    elif val < 0.001:\n",
    "        return star*3  # '***'\n",
    "    elif val < 0.01:\n",
    "        return star*2  # '**'\n",
    "    elif val < 0.05:\n",
    "        return star  # '*'\n",
    "    else:\n",
    "        return ''  # NONE and such\n",
    "\n",
    "def format_pvals(vals):\n",
    "    \"\"\"\n",
    "    Format p-values for display in a table or plot.\n",
    "    pval format:\n",
    "    Symbol  Meaning\n",
    "    ns      P > 0.05\n",
    "    *       P ≤ 0.05\n",
    "    **      P ≤ 0.01\n",
    "    ***     P ≤ 0.001\n",
    "    ****    P ≤ 0.0001 (For the last two choices only)\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    vals : np.array\n",
    "        p-values.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    formatted : np.array\n",
    "        Formatted p-values.\n",
    "    \"\"\"\n",
    "    \n",
    "    return np.vectorize(format_one_pval)(vals)\n",
    "\n",
    "def specify_genes(genes, source_species='mouse', target_species='human'):\n",
    "    \"\"\"Translates gene symbols from a species to another using orthologs.\"\"\"\n",
    "    genes = genes if isinstance(genes, list) else list(genes) if (isinstance(genes, np.ndarray) or isinstance(genes, pd.Index)) else [genes]\n",
    "    # load human-mouse orthology\n",
    "    tab = pd.read_csv('/Users/peidli/work/projects/PanCov19_Hamster/resources/homology_human_mouse/HMD_HumanPhenotype.rpt', sep='\\t',\n",
    "                  names=['human', 'human_gene_number (HGNC)', 'mouse', 'mouse_gene_number (MGI)', 'mouse protein'], usecols=[0,1,2,3,4])\n",
    "    human_mouse_orthology_dict = dict(zip(tab.human, tab.mouse))\n",
    "    mouse_human_orthology_dict = {key: value for (value, key) in human_mouse_orthology_dict.items()}\n",
    "    # translate genes\n",
    "    # TODO the mapping could lead to duplicate genes, is that a problem?\n",
    "    if source_species=='mouse' and target_species=='human':\n",
    "        return [mouse_human_orthology_dict[x] if x in mouse_human_orthology_dict.keys() else x.upper() for x in genes]\n",
    "    elif source_species=='human' and target_species=='mouse':\n",
    "        return [human_mouse_orthology_dict[x] if x in human_mouse_orthology_dict.keys() else x.capitalize() for x in genes]\n",
    "    elif target_species is 'mouse':\n",
    "        return [x.capitalize() for x in genes]\n",
    "    elif target_species is 'human':\n",
    "        return [x.upper() for x in genes]\n",
    "    else:\n",
    "        raise ValueError('Species '+target_species+' not known.')\n",
    "\n",
    "def score_genes(adata, gene_list, score_name, species='human', **kwargs):\n",
    "    \"\"\"Scores genes in gene_list and adds scores to adata.obs[score_name].\"\"\"\n",
    "    gene_list_ = specify_genes(gene_list, target_species=species)\n",
    "    sc.tl.score_genes(adata, gene_list_, score_name=score_name)\n",
    "\n",
    "def sanitize_name(s):\n",
    "    for rdict in [organism_rename_dict, time_rename_dict, dosage_rename_dict]:\n",
    "        for old, new in rdict.items():\n",
    "            s = s.replace(old, new)\n",
    "    return s.replace('_', ', ')\n",
    "\n",
    "### GSEA ###\n",
    "\n",
    "# Will annotate in plot significant gene sets containing any of these keywords\n",
    "keywords = ['infection', 'sars', 'covid', 'corona', 'innate', 'immune', 'cytokine', 'interferon', \n",
    "            'tgf', 'MTOR', 'Lung', 'neutro', 'apoptosis', 'hypoxia', 'nod-like' 'inflammation', 'inflammatory',\n",
    "           'DNA-sensing', 'calcium', 'influenza', 'hepatitis', 'jak', 'stat', 'il-', 'virus']\n",
    "\n",
    "def extract_gsea_results(prerank_object, gene_sets):\n",
    "    # prepare data\n",
    "    Y = prerank_object.res2d.copy()\n",
    "    \n",
    "    # clear names\n",
    "    Y['source']=None\n",
    "    Y['Enrichment_Term']=None\n",
    "    for index, row in Y.iterrows():\n",
    "        for gene_set in gene_sets:\n",
    "            if gene_set in row.Term:\n",
    "                Y.loc[index, 'source'] = gene_set\n",
    "                Y.loc[index, 'Enrichment_Term'] = row.Term.replace('__', '_').replace(gene_set+'_', '')\n",
    "    \n",
    "    Y = Y.sort_values('NES')\n",
    "    y = -np.log10(Y['FDR q-val'].astype(float))\n",
    "    Y['score'] = np.clip(y, 0, 5)# np.max(y[y!=np.infty]))\n",
    "    Y['significance_str'] = format_pvals(Y['FDR q-val'])\n",
    "    \n",
    "    # annotate immune related terms\n",
    "    Y['immune_related'] = False\n",
    "    for index, row in Y.iterrows():\n",
    "        if any([k.lower() in row.Term.lower() for k in keywords]):\n",
    "            Y.loc[index, 'immune_related'] = True\n",
    "    return Y\n",
    "\n",
    "def Uplot(prerank_object, gene_sets=[\"KEGG_2021_Human\", 'MSigDB_Hallmark_2020'], ax=None, legend=True):\n",
    "    ax = pl.subplots(figsize=(8,6))[-1] if ax is None else ax\n",
    "    \n",
    "    Y = extract_gsea_results(prerank_object, gene_sets)\n",
    "    \n",
    "    # annotate with arrow\n",
    "    x = Y[Y.Term=='MSigDB_Hallmark_2020__Inflammatory Response'].NES\n",
    "    y = Y[Y.Term=='MSigDB_Hallmark_2020__Inflammatory Response'].score\n",
    "    ax.annotate('Inflammatory\\nResponse', (x, y), xytext=(x+0.1, y-0.4), arrowprops=dict(arrowstyle=\"->\",\n",
    "                        connectionstyle=\"arc3\"), fontsize=5)\n",
    "    \n",
    "    sns.scatterplot(data=Y, x='NES', y='score', ax=ax, hue='immune_related', palette=['grey', 'red'], legend=legend)\n",
    "                    #hue='source', hue_order=np.sort(Y.source.unique()))\n",
    "    \n",
    "    ax.set_xlim(-2.5, 2.5)\n",
    "    # Move left y-axis and bottim x-axis to centre, passing through (0,0)\n",
    "    ax.spines['left'].set_position('center')\n",
    "    # Eliminate upper and right axes\n",
    "    ax.spines['right'].set_color('none')\n",
    "    ax.spines['top'].set_color('none')\n",
    "    \n",
    "    if legend:\n",
    "        ax.axhline(-np.log10(0.05), label=r'$\\alpha=0.05$')\n",
    "        ax.legend(title='GSEA\\nSignificance')#, bbox_to_anchor=(1,0.5))\n",
    "\n",
    "    ax.set_xlabel(f'GSEA Normalized Enrichment Score')\n",
    "    ax.set_ylabel('-Log10(q-value)', loc='top')\n",
    "    \n",
    "def do_gsea(datasets, celltype, key, load_key, gene_sets=[\"KEGG_2021_Human\", 'MSigDB_Hallmark_2020']):\n",
    "    if load_key==None:\n",
    "        if key=='severity':\n",
    "            load_key = 'severity'\n",
    "        elif key=='infection':\n",
    "            load_key = 'COVID19_vs_healthy'\n",
    "        else:\n",
    "            raise ValueError(f'Load_key for key {key} unknown. Please supply!')\n",
    "    \n",
    "    # load deseq results\n",
    "    dfs =[pd.read_csv(root / f'results/DESeq2/{key}/{celltype}_{d}_{load_key}.csv', index_col=0) for d in datasets]\n",
    "\n",
    "    # humanize hamster genes\n",
    "    for i, df_ in enumerate(dfs):\n",
    "        ds = datasets[i]\n",
    "        if 'hamster' in ds:\n",
    "            df_.index = specify_genes(df_.index, 'mouse', 'human')  # map to human using biomart\n",
    "            df_ = df_[~df_.index.duplicated(keep='first')]  # remove dups\n",
    "    \n",
    "    # GSEA for Pros:\n",
    "    ranked_genes = []\n",
    "    for i, df in enumerate(dfs):\n",
    "        # Step 1: Filter by mean expression (low expression makes statistics unreliable)\n",
    "        df = df[df.baseMean > 10]\n",
    "        # Step 2: Compute the neglog10 of p-value without multiple testing correction \n",
    "        # (we only want an ordering, the correction might lead to more draws in ranking)\n",
    "        df['score'] = - np.log10(df.pvalue)\n",
    "        # Step 3: Add sign from lfc\n",
    "        df['signed_score'] = df.score * np.sign(df.log2FoldChange)\n",
    "        # Step 4: Use that to rank genes\n",
    "        df = df[~df.signed_score.isna()].sort_values('signed_score', ascending=False)\n",
    "        ranked_genes.append(df[['signed_score']])\n",
    "    \n",
    "    # run GSEA\n",
    "    pre_res = []\n",
    "    for d, rank in zip(datasets, ranked_genes):\n",
    "        pre_res_ = gp.prerank(rnk=rank,\n",
    "                             gene_sets=gene_sets,\n",
    "                             threads=8,\n",
    "                             min_size=15,  # gene sets with less than that many genes in data are ignored\n",
    "                             max_size=1000,  # gene sets with more than that many genes in data are ignored\n",
    "                             permutation_num=1000,\n",
    "                             outdir=None, # don't write to disk\n",
    "                             seed=42,\n",
    "                             verbose=False\n",
    "                            )\n",
    "        pre_res.append(pre_res_)\n",
    "    return pre_res\n",
    "\n",
    "def DC_do_GSEA(df, gene_sets=['KEGG_2021_Human', 'MSigDB_Hallmark_2020']):\n",
    "    # humanize hamster genes\n",
    "    df.index = specify_genes(df.index, 'mouse', 'human')  # map to human using biomart\n",
    "    df = df[~df.index.duplicated(keep='first')]  # remove dups\n",
    "    df['signed_score'] = df.spearman_r\n",
    "    # Use that to rank genes\n",
    "    df = df[~df.signed_score.isna()].sort_values('signed_score', ascending=False)\n",
    "    rank = df[['signed_score']]\n",
    "\n",
    "    # run GSEA\n",
    "    pre_res_ = gp.prerank(rnk=rank,\n",
    "                            gene_sets=gene_sets,\n",
    "                            threads=8,\n",
    "                            min_size=15,  # gene sets with less than that many genes in data are ignored\n",
    "                            max_size=1000,  # gene sets with more than that many genes in data are ignored\n",
    "                            permutation_num=1000,\n",
    "                            outdir=None, # don't write to disk\n",
    "                            seed=42,\n",
    "                            verbose=False\n",
    "                        )\n",
    "    return pre_res_.res2d\n",
    "\n",
    "def plot_timecourse(adata, key, x='time', hue='organism', ci=99, markers=True, linewidth=2, ax=None, **kwargs):\n",
    "    keys = [key] if type(key)==str else key\n",
    "    N = len(keys)\n",
    "    axs = pl.subplots(1,N, figsize=[6*N,4], dpi=100)[-1] if ax is None else ax\n",
    "    axs = [axs] if N==1 else axs\n",
    "    for key, ax in zip(keys, axs):\n",
    "        df = extract_df(adata, [key, x, hue])\n",
    "        sns.lineplot(data=df, y=key, x=x, hue=hue, ci=ci, markers=markers, linewidth=linewidth, ax=ax, **kwargs)\n",
    "        ax.set_title(key)\n",
    "\n",
    "def plot_timecourse_dose(adata, key, x='time', ci=99, markers=True, linewidth=2, ax=None, **kwargs):\n",
    "    keys = [key] if type(key)==str else key\n",
    "    N = len(keys)\n",
    "    with sns.axes_style('whitegrid'):\n",
    "        axs = pl.subplots(1,N, figsize=[8*N,5], dpi=80)[-1] if ax is None else ax\n",
    "    axs = [axs] if N==1 else axs\n",
    "    for key, ax in zip(keys, axs):\n",
    "        df = extract_df(adata, [key, x, 'organism', 'dosage', 'organism_dosage'])\n",
    "        df_ = df[df.organism_dosage=='Dwarfhamster_no dose']\n",
    "        df = df[df.organism_dosage!='Dwarfhamster_no dose']\n",
    "        df_['organism_dosage'] = 'Dwarfhamster_low dose'\n",
    "        df = pd.concat([df, df_])\n",
    "        df_['organism_dosage'] = 'Dwarfhamster_high dose'\n",
    "        df = pd.concat([df, df_])\n",
    "        df.organism_dosage[df.organism_dosage.isin(['Goldhamster_no dose', 'Goldhamster_high dose'])] = 'Goldhamster'\n",
    "        df.index = np.sum(np.vstack([df.index, np.arange(len(df)).astype(str)]), axis=0)\n",
    "        sns.lineplot(data=df, y=key, x=x, hue='organism_dosage', ci=ci, markers=markers, linewidth=linewidth, ax=ax, **kwargs)\n",
    "        ax.set_title(key)\n",
    "\n",
    "tab = pd.read_csv(root / 'resources/biomart_mouse_gene_protein.csv')\n",
    "def annotate_genes(df, tab):\n",
    "    for gene in tqdm(df.index):\n",
    "        if gene in tab['Gene name'].values:\n",
    "            stab = tab[tab['Gene name']==gene]\n",
    "            for c in ['Gene stable ID', 'Protein stable ID', 'Gene type', 'Gene Synonym', 'Gene description']:\n",
    "                v = stab[c].unique()\n",
    "                v = v[~pd.isna(v)]\n",
    "                if len(v)==1:\n",
    "                    df.loc[gene, c] = v[0]\n",
    "                else:\n",
    "                    df.loc[gene, c] = ', '.join(v)\n",
    "    return df\n",
    "\n",
    "def lowess_with_confidence_bounds(\n",
    "    x, y, eval_x, N=200, conf_interval=0.95, lowess_kw=None\n",
    "):\n",
    "    \"\"\"\n",
    "    # https://www.statsmodels.org/dev/examples/notebooks/generated/lowess.html\n",
    "    Perform Lowess regression and determine a confidence interval by bootstrap resampling\n",
    "    \"\"\"\n",
    "    # Lowess smoothing\n",
    "    smoothed = sm.nonparametric.lowess(exog=x, endog=y, xvals=eval_x, **lowess_kw)\n",
    "\n",
    "    # Perform bootstrap resamplings of the data\n",
    "    # and  evaluate the smoothing at a fixed set of points\n",
    "    smoothed_values = np.empty((N, len(eval_x)))\n",
    "    for i in tqdm(range(N), leave=False):\n",
    "        sample = np.random.choice(len(x), len(x), replace=True)\n",
    "        sampled_x = x[sample]\n",
    "        sampled_y = y[sample]\n",
    "\n",
    "        smoothed_values[i] = sm.nonparametric.lowess(\n",
    "            exog=sampled_x, endog=sampled_y, xvals=eval_x, **lowess_kw\n",
    "        )\n",
    "\n",
    "    # Get the confidence interval\n",
    "    sorted_values = np.sort(smoothed_values, axis=0)\n",
    "    bound = int(N * (1 - conf_interval) / 2)\n",
    "    bottom = sorted_values[bound - 1]\n",
    "    top = sorted_values[-bound]\n",
    "\n",
    "    return smoothed, bottom, top\n",
    "\n",
    "def l2fc(x,y, pseudo_count=0.001):\n",
    "    \"\"\"Log2 fold change between two vectors x and y.\"\"\"\n",
    "    return np.log2(np.mean(x)+pseudo_count) - np.log2(np.mean(y)+pseudo_count)\n",
    "\n",
    "def mklist(str_or_list):\n",
    "    \"\"\"Make a list from a string or list.\"\"\"\n",
    "    return [str_or_list] if isinstance(str_or_list, str) else str_or_list\n",
    "\n",
    "def differential_test(adata, keys, groupby, replicate_key, control=None, correction_method='fdr_bh', sort_groups=True):\n",
    "    \"\"\"Perform differential test on keys (from adata.X or adata.obs) w.r.t. a grouping with specifyable control groups.\n",
    "\n",
    "    Uses t-test and by default corrects for multiple testing. Control groups can be specified for all groups or\n",
    "    as dict for every group separately.\n",
    "    Returns three pandas DataFrames: pval_adj_tab, l2fc_tab, fc_tab\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata: AnnData object\n",
    "        The AnnData object to extract data from.\n",
    "    keys: str or list of str\n",
    "        Keys to perform testing for.\n",
    "        Can be gene names or keys of adata.obs.\n",
    "    groupy: str\n",
    "        Key to categorical column in adata.obs.\n",
    "        Defines the groups for testing.\n",
    "    replicate_key: str\n",
    "        Key to categorical column in adata.obs.\n",
    "        Defines the replicates for testing.\n",
    "    control: str, list of str, or dict (default: `None`)\n",
    "        Defines the control group(s) to test against.\n",
    "        All values of control must be categories in adata.obs[groupby].\n",
    "        There are different ways of setting control:\n",
    "        - str: defines this group as control for all groups\n",
    "        - list of str: defines these groups as control for all groups\n",
    "        - dict: Dict must map every group of interest to at least one other group.\n",
    "    correction_method: str (default: `'fdr_bh'`)\n",
    "        Multiple testing correction method. Defaults to Benjamini-Hochberg.\n",
    "        Can be any viable value for parameter 'method'\n",
    "        statsmodels.stats.multitest.multipletests.\n",
    "    sort_groups: bool (default: `True`)\n",
    "        Wether to sort the groups, ignoring the categorical sorting in adata.obs[groupby].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrames\n",
    "        pval_adj_tab, l2fc_tab, fc_tab\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    groups = pd.unique(adata.obs[groupby])\n",
    "    groups = np.sort(groups) if sort_groups else groups\n",
    "    groups = list(pd.unique(list(control.keys()))\n",
    "                  ) if isinstance(control, dict) else groups\n",
    "\n",
    "    n_groups = len(groups)\n",
    "    if isinstance(keys, dict):\n",
    "        # convert dict values to list, ignore keys of dict.\n",
    "        keys = np.concatenate(list(keys.values()))\n",
    "    keys = mklist(keys)\n",
    "    not_found = [\n",
    "        key for key in keys if key not in adata.var_names and key not in adata.obs.keys()]\n",
    "    if len(not_found) > 0:\n",
    "        warn(f\"Could not find these keys: {not_found}\")\n",
    "        keys = [x for x in keys if x not in not_found]\n",
    "\n",
    "    pval_tab = pd.DataFrame(index=groups, columns=keys, dtype=float)\n",
    "    l2fc_tab = pd.DataFrame(index=groups, columns=keys, dtype=float)\n",
    "    fc_tab = pd.DataFrame(index=groups, columns=keys, dtype=float)\n",
    "\n",
    "    # Pseudobulked into replicates while recording groupby\n",
    "    X = extract_df(adata, keys + [replicate_key, groupby]).groupby(\n",
    "        [replicate_key, groupby]).mean().dropna().reset_index(level=1)\n",
    "    grouping = X[groupby]\n",
    "\n",
    "    for key in keys:\n",
    "        for g in groups:\n",
    "            x = X[grouping == g][key]\n",
    "            # determin control values\n",
    "            if control == None:\n",
    "                # test group against all other groups as control\n",
    "                xc = X[grouping != g][key]\n",
    "            elif isinstance(control, list) or control in groups:\n",
    "                # test between each group and the control group(s)\n",
    "                xc = X[np.isin(grouping, mklist(control))][key]\n",
    "            elif isinstance(control, dict):\n",
    "                # test between non-control group and control group, where control group if specifically given for each group\n",
    "                xc = X[np.isin(grouping, mklist(control[g]))][key]\n",
    "            else:\n",
    "                raise ValueError(\n",
    "                    f'Control group {control} was not found in groups {groups} given by groupby {groupby}.')\n",
    "\n",
    "            pval_tab.loc[g, key] = ttest_ind(x, xc)[1]\n",
    "            l2fc_tab.loc[g, key] = l2fc(x, xc)\n",
    "            fc_tab.loc[g, key] = np.mean(x) - np.mean(xc)\n",
    "\n",
    "    if correction_method != None:\n",
    "        # correct pvals\n",
    "        shape = pval_tab.values.shape\n",
    "        K = np.multiply(*shape)\n",
    "        mask = pval_tab.isnull()\n",
    "        pval_tab[mask] = 0\n",
    "        rejected, pvals_corrected, _, _ = multipletests(\n",
    "            pval_tab.values.reshape(K), alpha=0.05, method=correction_method)\n",
    "        pval_adj_tab = pd.DataFrame(np.array(pvals_corrected).reshape(\n",
    "            shape), index=groups, columns=keys)\n",
    "        pval_adj_tab[mask] = None\n",
    "    else:\n",
    "        pval_adj_tab = pval_tab\n",
    "    return pval_adj_tab, l2fc_tab, fc_tab\n",
    "\n",
    "def plot_differential_test(adata, keys, groupby, replicate_key, control=None, change='l2fc', correction_method='fdr_bh', \n",
    "                            sort_groups=True, kick_control=True, show=True,ax=None, center=0, robust=True, \n",
    "                            cluster=None, nan_color='lightgrey', swap_axes=False, annotate=True, **kwargs):\n",
    "    '''\n",
    "    Change may be 'fc' or 'l2fc'.\n",
    "    '''\n",
    "    pval_adj_tab, l2fc_tab, fc_tab = differential_test(\n",
    "        adata, keys, groupby, replicate_key, control, correction_method, sort_groups)\n",
    "    change_tab = eval(f'{change}_tab')\n",
    "    if kick_control and (isinstance(control, str) or isinstance(control, list)):\n",
    "        mask = [x for x in change_tab.index if x not in mklist(control)]\n",
    "        change_tab = change_tab.loc[mask]\n",
    "        pval_adj_tab = pval_adj_tab.loc[mask]\n",
    "    if cluster != None:\n",
    "        order = cluster_matrix(change_tab.fillna(\n",
    "            0), how=cluster, return_order=True, method='centroid')\n",
    "        change_tab = change_tab.iloc[order] if cluster == 'row' else change_tab.iloc[:, order]\n",
    "        pval_adj_tab = pval_adj_tab.iloc[order] if cluster == 'row' else pval_adj_tab.iloc[:, order]\n",
    "\n",
    "    if annotate:\n",
    "        cg = sns.heatmap(change_tab.T if swap_axes else change_tab,\n",
    "                         cmap='bwr', center=center, robust=robust,\n",
    "                         mask=pd.isna(pval_adj_tab.T) if swap_axes else pd.isna(\n",
    "                             pval_adj_tab),\n",
    "                         annot=format_pvals(pval_adj_tab.values.T) if swap_axes else format_pvals(pval_adj_tab.values), fmt='',\n",
    "                         ax=ax, **kwargs)\n",
    "    else:\n",
    "        cg = sns.heatmap(change_tab.T if swap_axes else change_tab,\n",
    "                         cmap='bwr', center=center, robust=robust,\n",
    "                         mask=pd.isna(pval_adj_tab.T) if swap_axes else pd.isna(\n",
    "                             pval_adj_tab),\n",
    "                         ax=ax, **kwargs)\n",
    "    if nan_color != None:\n",
    "        cg.set_facecolor(nan_color)\n",
    "    ax = pl.gca() if ax == None else ax\n",
    "    ax.set_title(f'{change} & adjusted pvalues\\nacross {groupby}')\n",
    "\n",
    "    # annotate columns if a dict is given\n",
    "    if type(keys) == dict:\n",
    "        if cluster == 'col':\n",
    "            # Obviously the cols need to stay grouped as in the dict, otherwise no brackets possible\n",
    "            warn(\n",
    "                'Clustering and column annotations can not be combined! Skipping annotation...')\n",
    "        else:\n",
    "            pos = 0\n",
    "            for key, value in keys.items():\n",
    "                # throw out genes not found\n",
    "                value = [x for x in value if x in change_tab.columns]\n",
    "                x = len(value) / 2 + pos - 0.5\n",
    "                width = len(value)\n",
    "                raise NotImplementedError('This is not finished yet!')\n",
    "                pos += len(value)\n",
    "    if show:\n",
    "        pl.show()\n",
    "    else:\n",
    "        return ax\n",
    "\n",
    "def get_foldchanges(adata, keys, groupby, control=None, log2=True, sort_groups=True):\n",
    "    \"\"\"Computes L2FCs or FCs of given keys (from adata.X or adata.obs) w.r.t. a grouping with specifyable control groups.\n",
    "    Control groups can be specified for all groups as str or\n",
    "    as dict for every group separately.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata: AnnData object\n",
    "        The AnnData object to extract data from.\n",
    "    keys: str or list of str\n",
    "        Keys to perform testing for.\n",
    "        Can be gene names or keys of adata.obs.\n",
    "    groupy: str\n",
    "        Key to categorical column in adata.obs.\n",
    "        Defines the groups for testing.\n",
    "    control: str, list of str, or dict (default: `None`)\n",
    "        Defines the control group(s) to compare against.\n",
    "        All values of control must be categories in adata.obs[groupby].\n",
    "        There are different ways of setting control:\n",
    "        - str: defines this group as control for all groups\n",
    "        - list of str: defines these groups as control for all groups\n",
    "        - dict: Dict must map every group of interest to at least one other group.\n",
    "    sort_groups: bool (default: `True`)\n",
    "        Wether to sort the groups, ignoring the categorical sorting in adata.obs[groupby].\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        l2fc_tab or fc_tab\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    groups = pd.unique(adata.obs[groupby])\n",
    "    groups = np.sort(groups) if sort_groups else groups\n",
    "    groups = list(pd.unique(list(control.keys()))) if isinstance(control, dict) else groups\n",
    "\n",
    "    n_groups = len(groups)\n",
    "    if isinstance(keys, dict):\n",
    "        # convert dict values to list, ignore keys of dict.\n",
    "        keys = np.sum(list(keys.values()))\n",
    "    keys = mklist(keys)\n",
    "    not_found = [key for key in keys if key not in adata.var_names and key not in adata.obs.keys()]\n",
    "    if len(not_found)>0:\n",
    "        warn(f\"Could not find these keys: {not_found}\")\n",
    "        keys = [x for x in keys if x not in not_found]\n",
    "    grouping = adata.obs[groupby]\n",
    "\n",
    "    fc_tab = pd.DataFrame(index=groups, columns=keys, dtype=float)\n",
    "    for key in keys:\n",
    "        # group-wise values\n",
    "        X = get_key_values(adata, key)\n",
    "\n",
    "        for g in groups:\n",
    "            x = X[grouping==g]\n",
    "            # determin control values\n",
    "            if control==None:\n",
    "                # comparison group against all other groups as control\n",
    "                xc = X[grouping!=g]\n",
    "            elif isinstance(control, list) or control in groups:\n",
    "                # comparison between each group and the control group(s)\n",
    "                xc = X[np.isin(grouping, mklist(control))]\n",
    "            elif isinstance(control, dict):\n",
    "                # comparison between non-control group and control group, where control group if specifically given for each group\n",
    "                xc = X[np.isin(grouping, mklist(control[g]))]\n",
    "            else:\n",
    "                raise ValueError(f'Control group {control} was not found in groups {groups} given by groupby {groupby}.')\n",
    "\n",
    "            fc_tab.loc[g, key] = l2fc(x, xc) if log2 else np.mean(x) - np.mean(xc)\n",
    "            \n",
    "    return fc_tab \n",
    "\n",
    "def plot_differential_test(adata, keys, groupby, replicate_key, control=None, change='l2fc', correction_method='fdr_bh', \n",
    "                            sort_groups=True, kick_control=True, show=True,ax=None, center=0, robust=True, \n",
    "                            cluster=None, nan_color='lightgrey', swap_axes=False, annotate=True, \n",
    "                            row_order=None,\n",
    "                            **kwargs):\n",
    "    '''\n",
    "    Change may be 'fc' or 'l2fc'.\n",
    "    '''\n",
    "    pval_adj_tab, l2fc_tab, fc_tab = differential_test(\n",
    "        adata, keys, groupby, replicate_key, control, correction_method, sort_groups)\n",
    "    change_tab = eval(f'{change}_tab')\n",
    "    if kick_control and (isinstance(control, str) or isinstance(control, list)):\n",
    "        mask = [x for x in change_tab.index if x not in mklist(control)]\n",
    "        change_tab = change_tab.loc[mask]\n",
    "        pval_adj_tab = pval_adj_tab.loc[mask]\n",
    "    if cluster != None:\n",
    "        order = cluster_matrix(change_tab.fillna(\n",
    "            0), how=cluster, return_order=True, method='centroid')\n",
    "        change_tab = change_tab.iloc[order] if cluster == 'row' else change_tab.iloc[:, order]\n",
    "        pval_adj_tab = pval_adj_tab.iloc[order] if cluster == 'row' else pval_adj_tab.iloc[:, order]\n",
    "    if row_order != None:\n",
    "        change_tab = change_tab.loc[row_order]\n",
    "        pval_adj_tab = pval_adj_tab.loc[row_order]\n",
    "        \n",
    "    if annotate:\n",
    "        cg = sns.heatmap(change_tab.T if swap_axes else change_tab,\n",
    "                         cmap='bwr', center=center, robust=robust,\n",
    "                         mask=pd.isna(pval_adj_tab.T) if swap_axes else pd.isna(\n",
    "                             pval_adj_tab),\n",
    "                         annot=format_pvals(pval_adj_tab.values.T) if swap_axes else format_pvals(pval_adj_tab.values), fmt='',\n",
    "                         ax=ax, **kwargs)\n",
    "    else:\n",
    "        cg = sns.heatmap(change_tab.T if swap_axes else change_tab,\n",
    "                         cmap='bwr', center=center, robust=robust,\n",
    "                         mask=pd.isna(pval_adj_tab.T) if swap_axes else pd.isna(\n",
    "                             pval_adj_tab),\n",
    "                         ax=ax, **kwargs)\n",
    "    if nan_color != None:\n",
    "        cg.set_facecolor(nan_color)\n",
    "    ax = pl.gca() if ax == None else ax\n",
    "    ax.set_title(f'{change} & adjusted pvalues\\nacross {groupby}')\n",
    "    if show:\n",
    "        pl.show()\n",
    "    else:\n",
    "        return ax\n",
    "\n",
    "\n",
    "def intersect_many(arrays):\n",
    "    \"\"\"Intersect many arrays, even more than two.\"\"\"\n",
    "    X = arrays[0]\n",
    "    for Y in arrays:\n",
    "        X = np.intersect1d(X,Y)\n",
    "    return X\n",
    "\n",
    "\n",
    "def shut_up_scanpy(func):\n",
    "    \"\"\"Tell scanpy to shut up.\"\"\"\n",
    "    @wraps(func)  # does these attribute copies commented out below...\n",
    "    def silent_wrapper(*args, **kwargs):\n",
    "        verbosity = sc.settings.verbosity.value\n",
    "        sc.settings.verbosity = 0\n",
    "        func(*args, **kwargs)\n",
    "        sc.settings.verbosity = verbosity\n",
    "    # silent_wrapper.__name__=func.__name__\n",
    "    # silent_wrapper.__annotations__=func.__annotations__\n",
    "    # silent_wrapper.__doc__=func.__doc__\n",
    "    # silent_wrapper.__defaults__=func.__defaults__\n",
    "    return silent_wrapper\n",
    "\n",
    "\n",
    "@shut_up_scanpy\n",
    "def score_hallmarks(adata, subset='all', signatures_path=signatures_path_, species='human'):\n",
    "    # subset can be a list of hallmarks, 'organoid' (), 'CRC' (~18) or 'all' (50 scores)\n",
    "    signatures_path = Path(signatures_path) if type(signatures_path) is str else signatures_path\n",
    "    tab = pd.read_csv(signatures_path / 'msigdb/hallmark/h.all.v6.2.symbols.gmt', sep='\\t', index_col=0, header=None).drop(1, axis=1).T\n",
    "    hallsigs={hallmark : tab[hallmark][~pd.isna(tab[hallmark])].values for hallmark in tab.columns}\n",
    "    if isinstance(subset, list):\n",
    "        selection = subset\n",
    "    elif subset == 'organoid':\n",
    "        selection = ['HALLMARK_DNA_REPAIR', 'HALLMARK_WNT_BETA_CATENIN_SIGNALING', 'HALLMARK_APOPTOSIS']\n",
    "    elif subset == 'CRC':  # TODO this list is bugged, some entries do not exist\n",
    "        selection = ['HALLMARK_DNA_REPAIR', 'HALLMARK_WNT_BETA_CATENIN_SIGNALING', 'HALLMARK_APOPTOSIS',\n",
    "        'HALLMARK_NOTCH_SIGNALING', 'HALLMARK_TNFA_SIGNALING_VIA_NFKB', 'HALLMARK_HYPOXIA', 'HALLMARK_TGF_BETA_SIGNALING',\n",
    "        'HALLMARK_MITOTIC_SPINDLE', 'HALLMARK_MTORC1_SIGNALING', 'HALLMARK_PI3K_AKT_MTOR_SIGNALING', 'HALLMARK_PROTEIN_SECRETION',\n",
    "        'HALLMARK_G2M_CHECKPOINT', 'HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION', 'HALLMARK_OXIDATIVE_PHOSPHORYLATION',\n",
    "        'HALLMARK_P53_PATHWAY', 'HALLMARK_ANGIOGENESIS', 'HALLMARK_KRAS_SIGNALING_UP', 'HALLMARK_KRAS_SIGNALING_DN',\n",
    "        'HALLMARK_GLYCOLYSIS']\n",
    "    elif subset == 'all':\n",
    "        selection = hallsigs.keys()\n",
    "    else:\n",
    "        raise ValueError('Please select a valid subset of hallmark to use. You can also choose \"all\".')\n",
    "    for hm in tqdm(selection):\n",
    "        score_genes(adata, hallsigs[hm], score_name=hm, species=species)\n",
    "\n",
    "def get_genefamily_percentage(adata, key='MT-', start=True, name='mito'):\n",
    "    keys = key if isinstance(key, list) else [key, '____ignore____']\n",
    "    if start:\n",
    "        family_genes = np.logical_or(*[adata.var_names.str.startswith(k) for k in keys])\n",
    "    else:\n",
    "        family_genes = np.logical_or(*[adata.var_names.str.endswith(k) for k in keys])\n",
    "    if issparse(adata.X):\n",
    "        adata.obs['percent_'+name] = np.sum(\n",
    "            adata[:, family_genes].X, axis=1).A1 * 100 / np.sum(adata.X, axis=1).A1\n",
    "    else:\n",
    "        adata.obs['percent_'+name] = np.sum(\n",
    "            adata[:, family_genes].X, axis=1) * 100 / np.sum(adata.X, axis=1)\n",
    "\n",
    "def get_mito_percentage(adata, species='human'):\n",
    "    key = 'MT-' if species == 'human' else 'mt-'\n",
    "    get_genefamily_percentage(adata, key=key, start=True, name='mito')\n",
    "\n",
    "def get_ribo_percentage(adata, species='human'):\n",
    "    key = specify_genes(['RPS', 'RPL'], target_species=species)\n",
    "    get_genefamily_percentage(adata, key=key, start=True, name='ribo')\n",
    "    \n",
    "def run_progeny(adata, species='human', top=100):\n",
    "    \"\"\"Run progeny on adata and add progeny scores to adata.obs\"\"\"\n",
    "    model = dc.get_progeny(organism=species, top=top)\n",
    "    dc.run_mlm(mat=adata, net=model, source='source', target='target', weight='weight', verbose=True, use_raw=False)\n",
    "    for key in adata.obsm['mlm_estimate'].columns: adata.obs[f'{key}_progeny'] = adata.obsm['mlm_estimate'][key]\n",
    "    \n",
    "def obs_obs_aggregate_heatmap(adata, obs_keys, groupby, standard_scale='var',\n",
    "        show=True, swap_axes=True, figsize=[10,3], obs_rotation=0, plot=True, **kwargs):\n",
    "    \"\"\"\n",
    "    Show a heatmap of pseudobulked (means) obs_keys grouped by groupby.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    data: :class:`~anndata.AnnData`\n",
    "        Annotated data matrix.\n",
    "    obs_keys : `list` or `np.ndarray` of numeric columns in adata.obs\n",
    "        Keys of observations to show (e.g. signatures or QC values).\n",
    "    groupby : `str` of categorical column in adata.obs\n",
    "        Key of observations grouping to consider.\n",
    "    standard_scale : `'var'`, `'group'` or `None`/`False` (default: `'var'`)\n",
    "        Axis to zscore normalization.\n",
    "    obs_rotation: float or int (default: 0)\n",
    "        Rotation of the obs_key labels.\n",
    "    show: bool (default: True)\n",
    "        Whether to call matplotlib.pyplot.show() in function call automatically.\n",
    "    plot: bool (default: True)\n",
    "        Whether to plot the heatmap.\n",
    "    Returns\n",
    "    -------\n",
    "    pandas DataFrame\n",
    "        DataFrame source for the heatmap.\n",
    "    \"\"\"\n",
    "    X = adata.obs.groupby(groupby).mean()[obs_keys]\n",
    "    X = X.T if swap_axes else X\n",
    "    Z = zscore(X.values, axis=1) if standard_scale == 'var' else zscore(X.values, axis=0) if standard_scale == 'group' else X.values\n",
    "    X = pd.DataFrame(Z, index=X.index, columns=X.columns)\n",
    "    if plot:\n",
    "        pl.figure(figsize=figsize)\n",
    "        sns.heatmap(X, **kwargs)\n",
    "        pl.xlabel(groupby)\n",
    "        pl.yticks(rotation=obs_rotation) if swap_axes else pl.xticks(rotation=obs_rotation)\n",
    "        if show: pl.show()\n",
    "    return X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5a7f66da",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path('/Users/peidli/work/data/COVID')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3aa98c46",
   "metadata": {},
   "source": [
    "# Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e7a30ceb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-16T10:41:51.521943Z",
     "start_time": "2023-01-16T10:30:30.807857Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "superdata = sc.read(data_path / \"Pancov19_prepd_all.h5\")\n",
    "\n",
    "# Import color options\n",
    "import yaml\n",
    "with open(root / 'config/config.yaml', 'r') as file:\n",
    "    config = yaml.safe_load(file)\n",
    "celltype_colors = config['celltype_colors']\n",
    "celltype_rough_colors = config['celltype_rough_colors']\n",
    "organism_colors = config['organism_colors']\n",
    "organism_dosage_colors = config['organism_dosage_colors']\n",
    "dosage_colors = config['dosage_colors']\n",
    "\n",
    "# Newer names, more accurate\n",
    "organism_rename_dict = {\n",
    "    'Dwarfhamster': 'Roborovski hamster',\n",
    "    'Goldhamster': 'Syrian hamster'\n",
    "}\n",
    "time_rename_dict = {\n",
    "    'D0': '0 dpi',\n",
    "    'D2': '2 dpi',\n",
    "    'D3': '3 dpi',\n",
    "    'D5': '5 dpi',\n",
    "    'E14': '14 dpi'\n",
    "}\n",
    "dosage_rename_dict = {\n",
    "    'no dose': '0 pfu',\n",
    "    'low dose': r'$1\\times10^4$ pfu',\n",
    "    'high dose': r'$1\\times10^5$ pfu'\n",
    "}\n",
    "\n",
    "def sanitize_name(s):\n",
    "    for rdict in [organism_rename_dict, time_rename_dict, dosage_rename_dict]:\n",
    "        for old, new in rdict.items():\n",
    "            s = s.replace(old, new)\n",
    "            \n",
    "    return s.replace('_', ', ')\n",
    "\n",
    "from matplotlib.cm import get_cmap\n",
    "# not super pretty, but ok for now\n",
    "time_colors = {\n",
    "    'D0': 'grey',\n",
    "    'D2': get_cmap('viridis_r')(0),\n",
    "    'D3': get_cmap('viridis_r')(1/3),\n",
    "    'D5': get_cmap('viridis_r')(2/3),\n",
    "    'E14': get_cmap('viridis_r')(1),\n",
    "              }\n",
    "    \n",
    "# Translate color palettes to new names as well\n",
    "Organism_colors = {sanitize_name(key) : value for key, value in config['organism_colors'].items()}\n",
    "Organism_Dosage_colors = {sanitize_name(key) : value for key, value in config['organism_dosage_colors'].items()}\n",
    "Dosage_colors = {sanitize_name(key) : value for key, value in config['dosage_colors'].items()}\n",
    "Time_colors = {sanitize_name(key) : value for key, value in time_colors.items()}\n",
    "\n",
    "# define control dicts for DE testing etc.\n",
    "control_dict = {x: ['pr_D0_Z1_L','pr_D0_Z2_L','pr_D0_Z3_L'] if 'pr' in x else ['ma_d0_lung_1', 'ma_d0_lung_2', 'ma_d0_lung_3'] for x in pd.unique(superdata.obs['orig.ident']) if '0' not in x}\n",
    "control_dict_odt = {x: 'Dwarfhamster_no dose_D0' if 'Dwarfhamster' in x else 'Goldhamster_no dose_D0' for x in superdata.obs.organism_dosage_time.unique() if 'D0' not in x}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "dee26a39",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "color key is invalid! pass valid observation annotation or a gene name",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[21], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m genes \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSidt\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSelp\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m----> 2\u001b[0m \u001b[43mscv\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscatter\u001b[49m\u001b[43m(\u001b[49m\u001b[43msuperdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbasis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mumap\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgenes\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/sc_env/lib/python3.9/site-packages/scvelo/plotting/scatter.py:237\u001b[0m, in \u001b[0;36mscatter\u001b[0;34m(adata, basis, x, y, vkey, color, use_raw, layer, color_map, colorbar, palette, size, alpha, linewidth, linecolor, perc, groups, sort_order, components, projection, legend_loc, legend_loc_lines, legend_fontsize, legend_fontweight, legend_fontoutline, legend_align_text, xlabel, ylabel, title, fontsize, figsize, xlim, ylim, add_density, add_assignments, add_linfit, add_polyfit, add_rug, add_text, add_text_pos, add_margin, add_outline, outline_width, outline_color, n_convolve, smooth, normalize_data, rescale_color, color_gradients, dpi, frameon, zorder, ncols, nrows, wspace, hspace, show, save, ax, **kwargs)\u001b[0m\n\u001b[1;32m    232\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m mkeys:  \u001b[38;5;66;03m# multi_kwargs[key] = key[i] if is multikey else key\u001b[39;00m\n\u001b[1;32m    233\u001b[0m             multi_kwargs[key] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28meval\u001b[39m(\n\u001b[1;32m    234\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;124m[i * (len(\u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;124m) > i)] if is_list(\u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;124m) else \u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(key)\n\u001b[1;32m    235\u001b[0m             )\n\u001b[1;32m    236\u001b[0m         ax\u001b[38;5;241m.\u001b[39mappend(\n\u001b[0;32m--> 237\u001b[0m             \u001b[43mscatter\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    238\u001b[0m \u001b[43m                \u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    239\u001b[0m \u001b[43m                \u001b[49m\u001b[43max\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msubplot\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    240\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmulti_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    241\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mscatter_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    242\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    243\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    244\u001b[0m         )\n\u001b[1;32m    246\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m frameon \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(ylabel, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    247\u001b[0m     set_label(xlabel, ylabel, fontsize, ax\u001b[38;5;241m=\u001b[39max[\u001b[38;5;241m0\u001b[39m], fontweight\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbold\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/mambaforge/envs/sc_env/lib/python3.9/site-packages/scvelo/plotting/scatter.py:534\u001b[0m, in \u001b[0;36mscatter\u001b[0;34m(adata, basis, x, y, vkey, color, use_raw, layer, color_map, colorbar, palette, size, alpha, linewidth, linecolor, perc, groups, sort_order, components, projection, legend_loc, legend_loc_lines, legend_fontsize, legend_fontweight, legend_fontoutline, legend_align_text, xlabel, ylabel, title, fontsize, figsize, xlim, ylim, add_density, add_assignments, add_linfit, add_polyfit, add_rug, add_text, add_text_pos, add_margin, add_outline, outline_width, outline_color, n_convolve, smooth, normalize_data, rescale_color, color_gradients, dpi, frameon, zorder, ncols, nrows, wspace, hspace, show, save, ax, **kwargs)\u001b[0m\n\u001b[1;32m    531\u001b[0m     c \u001b[38;5;241m=\u001b[39m interpret_colorkey(adata, basis, color, perc, use_raw)\n\u001b[1;32m    532\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    533\u001b[0m     \u001b[38;5;66;03m# embedding, gene trend etc.\u001b[39;00m\n\u001b[0;32m--> 534\u001b[0m     c \u001b[38;5;241m=\u001b[39m \u001b[43minterpret_colorkey\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlayer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mperc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muse_raw\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    536\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m c \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(c, \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(c[\u001b[38;5;241m0\u001b[39m], \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    537\u001b[0m     \u001b[38;5;66;03m# smooth color values across neighbors and rescale\u001b[39;00m\n\u001b[1;32m    538\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m smooth \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(c) \u001b[38;5;241m==\u001b[39m adata\u001b[38;5;241m.\u001b[39mn_obs:\n",
      "File \u001b[0;32m~/mambaforge/envs/sc_env/lib/python3.9/site-packages/scvelo/plotting/utils.py:810\u001b[0m, in \u001b[0;36minterpret_colorkey\u001b[0;34m(adata, c, layer, perc, use_raw)\u001b[0m\n\u001b[1;32m    808\u001b[0m     c \u001b[38;5;241m=\u001b[39m obs\u001b[38;5;241m.\u001b[39mastype(np\u001b[38;5;241m.\u001b[39mfloat32)\u001b[38;5;241m.\u001b[39meval(c)\n\u001b[1;32m    809\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_color_like(c):\n\u001b[0;32m--> 810\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    811\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcolor key is invalid! pass valid observation annotation or a gene name\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    812\u001b[0m     )\n\u001b[1;32m    813\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(c, \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m perc \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    814\u001b[0m     c \u001b[38;5;241m=\u001b[39m clip(c, perc\u001b[38;5;241m=\u001b[39mperc)\n",
      "\u001b[0;31mValueError\u001b[0m: color key is invalid! pass valid observation annotation or a gene name"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAf0AAAGfCAYAAABV8AvZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAZI0lEQVR4nO3db2yV5f348Q9QKet/ESchLRMmIKU4E2SMBJ2bWcTpFIOjgrLF8GCOsKVRoyPYMSSThcygxi3KA4dOiGCZzCgzC4uV7I+pmYtREBHHhMEAB6EtdlYr9+8R3bejSG+o5Veu1ys5D86Vq3c/vdL0zTn09AzIsiwLAOCsN/BMDwAA9A3RB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQiFOO/vvvvx8XXXRRNDY2nnDPxo0bY+LEiVFcXBzjx4+P559//lQ/HQBwmk4p+n/6059i6tSp8e67755wzzvvvBMzZ86MpUuXRnNzcyxZsiRmzZoVe/bsOeVhAYBTlzv6TzzxRMyZMyd++tOfnnTf5ZdfHjNmzIiCgoKYNWtWfPWrX42VK1ee8rAAwKkryPsBV199ddxyyy1RUFAQN9988wn3bdmyJSZOnNhlrbq6Ol5//fVu97e3t0d7e3vn/aNHj8ahQ4fivPPOiwEDBuQdEwD6vSzLorW1NUaMGBEDB57+r+Hljv7w4cN7tK+1tTWKi4u7rBUVFcWRI0e63b9s2bJYsmRJ3nEA4Ky3e/fuqKysPO3r5I5+TxUXF0dbW1uXtba2tigtLe12/8KFC+OOO+7ovN/c3BwjR46M3bt3R1lZ2Wc1JgD8f6ulpSWqqqpO2M68PrPo19TUxGuvvdZlbevWrXHZZZd1u7+wsDAKCwuPWy8rKxN9AJLWW//N/Zm9Tn/u3LnR2NgY69ati46Ojli3bl00NjbG3LlzP6tPCQB8il6NfklJSaxevToiIi6++OLYsGFD3H///XHuuefGfffdF+vXr4+xY8f25qcEAHpoQJZl2ZkeojstLS1RXl4ezc3Nnt4HIEm93UJ/hhcAEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAInJH/8CBAzFjxoyoqKiIYcOGRV1dXXR0dHS796GHHopRo0ZFWVlZXHLJJbF+/frTHhgAODW5o19bWxslJSWxd+/eaGpqik2bNsWKFSuO2/e73/0u7r///njxxRejpaUlFi9eHLNmzYp//OMfvTE3AJBTrujv2LEjGhsbY/ny5VFUVBSjR4+O+vr6eOSRR47b+9Zbb0WWZXH06NHIsiwGDRoUgwcPjoKCgl4bHgDouVwF3rJlSwwdOjRGjBjRuVZdXR27du2Kw4cPR0VFRef67Nmz41e/+lVUV1fHoEGDYsCAAfHUU09FZWVlt9dub2+P9vb2zvstLS05vxQA4NPkeqTf2toaxcXFXdaKiooiIuLIkSNd1j/66KO49NJLo6mpKdra2mLlypUxb968eOONN7q99rJly6K8vLzzVlVVlWc0AOAkckW/uLg42trauqwdu19aWtplfcGCBTFhwoSYPHlyDB48OG677baYOnVqrFq1qttrL1y4MJqbmztvu3fvzjMaAHASuaJfU1MTBw8ejP3793eubd26NSorK6O8vLzL3l27dnV5uj4i4pxzzonBgwd3e+3CwsIoKyvrcgMAek+u6I8ZMyamTZsWdXV10draGjt37oylS5fGvHnzjtt7/fXXxyOPPBKvvfZaHD16NBoaGuKll16K2traXhseAOi53L9K39DQEAsWLIhRo0bFwIED4zvf+U7U19dHRERJSUk89thjccstt8TixYtj0KBBMXPmzDh06FCMGTMmNmzYEJdeemlvfw0AQA8MyLIsO9NDdKelpSXKy8ujubnZU/0AJKm3W+jP8AJAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIRO7oHzhwIGbMmBEVFRUxbNiwqKuri46Ojm73vvzyyzFlypQoKSmJqqqqWLZs2WkPDACcmtzRr62tjZKSkti7d280NTXFpk2bYsWKFcft27ZtW3zzm9+M+fPnR2tra7zwwgvxwAMPRENDQ68MDgDkMyDLsqynm3fs2BFjxoyJPXv2xIgRIyIiYu3atXH33XfHe++912XvD37wgzh06FCsXr26c2379u1RVlYWw4cPP+nnamlpifLy8mhubo6ysrKejggAZ43ebmGuR/pbtmyJoUOHdgY/IqK6ujp27doVhw8f7rK3qakpLrzwwpg9e3YMGzYsxo8fH42NjScMfnt7e7S0tHS5AQC9J1f0W1tbo7i4uMtaUVFRREQcOXKky/qhQ4fi4YcfjltvvTX27dsXjz32WNx1110nfHp/2bJlUV5e3nmrqqrKMxoAcBK5ol9cXBxtbW1d1o7dLy0t7bJeWFgYN9xwQ1x77bVRUFAQV1xxRcydOzfWrVvX7bUXLlwYzc3Nnbfdu3fnGQ0AOImCPJtramri4MGDsX///rjgggsiImLr1q1RWVkZ5eXlXfZWV1dHe3t7l7VPPvkkTvQrBIWFhVFYWJhnHAAgh1yP9MeMGRPTpk2Lurq6aG1tjZ07d8bSpUtj3rx5x+29/fbbY8OGDfHUU09FlmWxefPmWL16dcydO7fXhgcAei73S/YaGhqio6MjRo0aFVOmTInp06dHfX19RESUlJR0/rb+17/+9XjuuefioYceivLy8rjtttvi5z//eVx//fW9+xUAAD2S6yV7fclL9gBI3Rl9yR4A0H+JPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkIjc0T9w4EDMmDEjKioqYtiwYVFXVxcdHR2f+jFvvvlmFBUVRWNj46nOCQCcptzRr62tjZKSkti7d280NTXFpk2bYsWKFSfc39bWFrNnz47//Oc/pzUoAHB6ckV/x44d0djYGMuXL4+ioqIYPXp01NfXxyOPPHLCj5k/f37ceOONJ712e3t7tLS0dLkBAL0nV/S3bNkSQ4cOjREjRnSuVVdXx65du+Lw4cPH7X/yySdjx44dsXjx4pNee9myZVFeXt55q6qqyjMaAHASuaLf2toaxcXFXdaKiooiIuLIkSNd1rdt2xaLFi2KNWvWxKBBg0567YULF0Zzc3Pnbffu3XlGAwBOoiDP5uLi4mhra+uydux+aWlp59qHH34YtbW18eCDD8bIkSN7dO3CwsIoLCzMMw4AkEOuR/o1NTVx8ODB2L9/f+fa1q1bo7KyMsrLyzvXXn311di+fXvMmzcvKioqoqKiIiIirrvuupg/f37vTA4A5DIgy7IszwdcfvnlUVlZGStXrox///vf8a1vfStuuumm+MlPfvLpn2jAgHjppZfiyiuv7NHnaWlpifLy8mhubo6ysrI8IwLAWaG3W5j7JXsNDQ3R0dERo0aNiilTpsT06dOjvr4+IiJKSkpi9erVpz0UAND7cj/S7yse6QOQujP+SB8A6J9EHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASETu6B84cCBmzJgRFRUVMWzYsKirq4uOjo5u9z766KMxbty4KC0tjbFjx8Yvf/nL0x4YADg1uaNfW1sbJSUlsXfv3mhqaopNmzbFihUrjtu3YcOGWLhwYTzxxBPR0tISTzzxRCxatCjWr1/fK4MDAPnkiv6OHTuisbExli9fHkVFRTF69Oior6+PRx555Li9e/fujR/96Efxla98JQYMGBBTp06Nr33ta7F58+Zur93e3h4tLS1dbgBA78kV/S1btsTQoUNjxIgRnWvV1dWxa9euOHz4cJe98+fPj3vuuafz/oEDB2Lz5s0xadKkbq+9bNmyKC8v77xVVVXlGQ0AOIlc0W9tbY3i4uIua0VFRRERceTIkRN+3L59++Kaa66JSZMmxZw5c7rds3Dhwmhubu687d69O89oAMBJ5Ip+cXFxtLW1dVk7dr+0tLTbj3nllVdi8uTJMW7cuHjuueeioKCg232FhYVRVlbW5QYA9J5c0a+pqYmDBw/G/v37O9e2bt0alZWVUV5eftz+xx9/PK666qqoq6uLNWvWRGFh4elPDACcklzRHzNmTEybNi3q6uqitbU1du7cGUuXLo158+Ydt3f9+vXx/e9/P37zm9/EnXfe2WsDAwCnJvdL9hoaGqKjoyNGjRoVU6ZMienTp0d9fX1ERJSUlMTq1asjImLJkiXR0dERM2fOjJKSks7b7bff3rtfAQDQIwOyLMvO9BDdaWlpifLy8mhubvb/+wAkqbdb6M/wAkAiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEiE6ANAIkQfABIh+gCQCNEHgESIPgAkQvQBIBGiDwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCJEHwASIfoAkAjRB4BEiD4AJEL0ASARog8AiRB9AEhE7ugfOHAgZsyYERUVFTFs2LCoq6uLjo6Obvdu3LgxJk6cGMXFxTF+/Ph4/vnnT3tgAODU5I5+bW1tlJSUxN69e6OpqSk2bdoUK1asOG7fO++8EzNnzoylS5dGc3NzLFmyJGbNmhV79uzplcEBgHwGZFmW9XTzjh07YsyYMbFnz54YMWJERESsXbs27r777njvvfe67L333nujqakpfv/733euXXPNNfHlL385lixZcty129vbo729vfN+c3NzjBw5Mnbv3h1lZWW5vzAA6O9aWlqiqqoqDh8+HOXl5ad9vYI8m7ds2RJDhw7tDH5ERHV1dezatSsOHz4cFRUVXfZOnDixy8dXV1fH66+/3u21ly1b1u0/BqqqqvKMCABnnYMHD/Z99FtbW6O4uLjLWlFRUUREHDlypEv0T7T3yJEj3V574cKFcccdd3TeP3z4cHzhC1+IXbt29coXSveO/SvSMyqfLefcd5x133DOfePYs95Dhw7tlevlin5xcXG0tbV1WTt2v7S0tEd7/3ffMYWFhVFYWHjcenl5uW+oPlBWVuac+4Bz7jvOum84574xcGDvvNgu11Vqamri4MGDsX///s61rVu3RmVl5XGPxmtqamLLli1d1rZu3Ro1NTWnMS4AcKpyRX/MmDExbdq0qKuri9bW1ti5c2csXbo05s2bd9zeuXPnRmNjY6xbty46Ojpi3bp10djYGHPnzu214QGAnsv9fEFDQ0N0dHTEqFGjYsqUKTF9+vSor6+PiIiSkpJYvXp1RERcfPHFsWHDhrj//vvj3HPPjfvuuy/Wr18fY8eO7dHnKSwsjMWLF3f7lD+9xzn3Defcd5x133DOfaO3zznXS/YAgP7Ln+EFgESIPgAkQvQBIBGiDwCJOKPR9459fSPPOT/66KMxbty4KC0tjbFjx8Yvf/nLPp62/8pzzse8+eabUVRUFI2NjX0z5Fkiz1m//PLLMWXKlCgpKYmqqqpYtmxZH0/bf+U554ceeihGjRoVZWVlcckll8T69ev7eNr+7/3334+LLrroU38enHYLszPoyiuvzG655Zbsgw8+yN59991swoQJ2fLly4/bt3379mzIkCHZs88+m3388cfZ2rVrs8997nPZP//5zzMwdf/T03N+9tlns4qKiuwvf/lLdvTo0ezPf/5zVlFRkTU0NJyBqfufnp7zMR988EFWU1OTRUT20ksv9d2gZ4GenvVbb72VFRUVZatWrcqOHj2avf7669l5552XPfPMM2dg6v6np+e8cePG7POf/3y2bdu2LMuyrKGhIRs4cGC2c+fOPp64//rjH/+YffGLX/zUnwe90cIzFv133nkni4hsz549nWtPP/10NnLkyOP2Llq0KPvGN77RZW369OnZj3/84898zv4uzzn/4he/yH72s591WbvxxhuzH/7wh5/5nP1dnnM+5rvf/W5WX18v+jnlOesFCxZkc+bM6bL29ttvZ//6178+8zn7uzzn/MADD2Tnn39+tnXr1uzo0aPZs88+mw0ZMiTbvXt3X47cb61atSobOXJk9vTTT3/qz4PeaOEZe3r/ZO/Y979787xjH/+V55znz58f99xzT+f9AwcOxObNm2PSpEl9NW6/leecIyKefPLJ2LFjRyxevLgPpzw75DnrpqamuPDCC2P27NkxbNiwGD9+fDQ2Nsbw4cP7eOr+J885z549Oy644IKorq6Oc845J7797W/HqlWrorKyso+n7p+uvvrqePfdd6O2tvZT9/VGC89Y9E/2jn092Xuid+zjv/Kc8/+1b9++uOaaa2LSpEkxZ86cz3TGs0Gec962bVssWrQo1qxZE4MGDeqzGc8Wec760KFD8fDDD8ett94a+/bti8ceeyzuuuuuaGho6LN5+6s85/zRRx/FpZdeGk1NTdHW1hYrV66MefPmxRtvvNFn8/Znw4cPj4KCk7//XW+08IxF/7N8xz7+K885H/PKK6/E5MmTY9y4cfHcc8/16JsxdT095w8//DBqa2vjwQcfjJEjR/bpjGeLPN/ThYWFccMNN8S1114bBQUFccUVV8TcuXNj3bp1fTZvf5XnnBcsWBATJkyIyZMnx+DBg+O2226LqVOnxqpVq/pq3CT0RgvPWPS9Y1/fyHPOERGPP/54XHXVVVFXVxdr1qzxd7V7qKfn/Oqrr8b27dtj3rx5UVFRERUVFRERcd1118X8+fP7eux+Kc/3dHV1dbS3t3dZ++STTyLz18dPKs8579q167hzPuecc2Lw4MF9MmsqeqWFp/G7B6dt2rRp2c0335y1tLRkf//737MJEyZkixcvPm7fW2+9lQ0ZMiRbu3Zt528sDhkyJHv77bf7fuh+qKfn3NDQkA0ePDh78cUX+37Is0BPz/l/hV/ky62nZ/2HP/whKygoyH79619nR48ezV5++eWspKQk++1vf9v3Q/dDPT3ne++9Nxs2bFj217/+Nfvkk0+yZ555JhsyZEj2t7/9rc9n7u8+7edBb7TwjEZ/37592U033ZSdd9552fnnn5/deeedWUdHR5ZlWVZcXJw99dRTnXtffPHF7Etf+lJWUlKSTZgwIXvhhRfO1Nj9Tk/PeeLEidnAgQOz4uLiLrfvfe97Z3L8fiPP9/P/Jfr55TnrjRs3ZpdddllWWlqajR49Onv00UfP1Nj9Tk/P+eOPP84WL16cXXjhhVlZWVk2adIkDx5O0f/+POjtFnqXPQBIhD/DCwCJEH0ASIToA0AiRB8AEiH6AJAI0QeARIg+ACRC9AEgEaIPAIkQfQBIhOgDQCL+H7rHTpwl5WJOAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1280x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes = ['Sidt', 'Selp']\n",
    "scv.pl.scatter(superdata, basis='umap', color=genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0f09ee84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sidt1 in var_names:\t True\n",
      "Sidt2 in var_names:\t True\n",
      "Selplg in var_names:\t True\n",
      "Selp in var_names:\t False\n"
     ]
    }
   ],
   "source": [
    "genes = ['Sidt1', 'Sidt2', 'Selplg', 'Selp']\n",
    "for gene in genes:\n",
    "    print(gene, 'in var_names:\\t', gene in superdata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1cf0989e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sel1l', 'Sel1l3', 'Sele', 'Selenbp2', 'Selenof', 'Selenoh', 'Selenoi',\n",
       "       'Selenok', 'Selenom', 'Selenon', 'Selenoo', 'Selenop', 'Selenos',\n",
       "       'Selenot', 'Selenow', 'Sell', 'Selplg'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "superdata.var_names[superdata.var_names.str.startswith('Sel')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a5a462",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "sc_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "vscode": {
   "interpreter": {
    "hash": "b286f6d81d26b8c5e9b75f0b330b147d4e82d310a48f1d10d555cf8099dd470e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
